common/src/androidTest/java/com/microsoft/identity/common/crypto/AndroidWrappedKeyLoaderTest.java

```diff
@@ -0,0 +1,178 @@
+        AndroidKeyStoreUtil.deleteKey(MOCK_KEY_ALIAS);
+
+        // Read again - should be empty.
+        Assert.assertNull(AndroidKeyStoreUtil.readKey(MOCK_KEY_ALIAS));
+    }
+
+    private AlgorithmParameterSpec getMockKeyPairGeneratorSpec(final String alias){
+        final Date startDate = new Date();
+        final Date endDate =  new Date(startDate.getTime() + 1000000);
+
+        return new KeyPairGeneratorSpec.Builder(context)
+                .setAlias(alias)
+                .setSubject(new X500Principal("CN=SOME_CN, OU=SOME_OU"))
+                .setSerialNumber(BigInteger.ONE)
+                .setStartDate(startDate)
+                .setEndDate(endDate)
+                .build();
+    }
+
+    @Test
+    public void testGenerateKey() throws ClientException {
+        final AndroidWrappedKeyLoader keyLoader = new AndroidWrappedKeyLoader(MOCK_KEY_ALIAS, context, null);
+        final SecretKey secretKey = keyLoader.generateRandomKey();
+
+        Assert.assertEquals(AES256KeyLoader.AES_ALGORITHM, secretKey.getAlgorithm());
+    }
+
+    @Test
+    public void testReadKeyDirectly() throws ClientException {
+        final AndroidWrappedKeyLoader keyLoader = initKeyLoaderWithKeyEntry();
+        final SecretKey secretKey = keyLoader.getKey();
+        final SecretKey storedSecretKey = keyLoader.readSecretKeyFromStorage();
+
+        // They're not the same object!
+        Assert.assertNotSame(secretKey, storedSecretKey);
+
+        Assert.assertEquals(AES256KeyLoader.AES_ALGORITHM, secretKey.getAlgorithm());
+        Assert.assertEquals(AES256KeyLoader.AES_ALGORITHM , storedSecretKey.getAlgorithm());
+
+        Assert.assertArrayEquals(secretKey.getEncoded(), storedSecretKey.getEncoded());
+        Assert.assertEquals(secretKey.getFormat(), storedSecretKey.getFormat());
+    }
+
+    @Test
+    public void testLoadKey() throws ClientException {
+        // Nothing exists. This load key function should generate a key if the key hasn't exist.
+        Assert.assertNull(AndroidKeyStoreUtil.readKey(MOCK_KEY_ALIAS));
+        Assert.assertNull(FileUtil.readFromFile(getKeyFile(), AndroidWrappedKeyLoader.KEY_FILE_SIZE));
+
+        final AndroidWrappedKeyLoader keyLoader = new AndroidWrappedKeyLoader(MOCK_KEY_ALIAS, context, null);
+        final SecretKey secretKey = keyLoader.getKey();
+        Assert.assertNotNull(keyLoader.getCachedKey());
+
+        Assert.assertEquals(AES256KeyLoader.AES_ALGORITHM, secretKey.getAlgorithm());
+        Assert.assertArrayEquals(secretKey.getEncoded(), keyLoader.getCachedKey().getEncoded());
+        Assert.assertEquals(secretKey.getFormat(), keyLoader.getCachedKey().getFormat());
+    }
+
+    @Test
+    public void testLoadDeletedKeyStoreKey() throws ClientException {
+        final AndroidWrappedKeyLoader keyLoader = initKeyLoaderWithKeyEntry();
+
+        AndroidKeyStoreUtil.deleteKey(MOCK_KEY_ALIAS);
+        // Cached key should not be wiped - yet, since we delete directly in keychain.
```

Comment:
Do we want that behavior, or is it essentially a bug?

Reply:
It's a bug, but after a second though, i think this (and the case below) could be improved.

Reply:
Essentially, this boils down to the question - "Should we cache the key"?

These test cases simulate the scenarios where the key file or the keychain is corrupted (or suddenly gone), and this layer wouldn't be aware.

I think i'll keep the same behavior for now, but will leave comment that this could be improved.

Reply:
Per 100 get key operation
/w cached key = 3s
without = 25s

Reply:
Well, you could be aware if we watched the file.

Reply:
Tested. Checking keystore and file is cheap (no difference in s) - i can add this.

Reply:
(Need to singleton-ize the Keystore object though).